### PR TITLE
Added new functionality: support for iterating across parameters in simulations

### PR DESCRIPTION
**Pull request by ArcadeShrimp**
param_iter()
_param_generator()

It may be useful to iterate over specific parameters when generating synthetic power spectra. I have added functions (mentioned above) to do such tasks. Responds to #96

### CODE REVIEW

File: fooof/synth.py [hunk 1]
```diff
@@ -17,6 +19,69 @@
...
+            length = len(params[i])
+            num_iters += 1
+            ind = i
+    
+        if num_iters > 1:
+            raise Warning("Iteration is only supported on one parameter")
+            
+    temp = params
+    c = ( _param_generator(params,ind,x) for x in params[ind])     
+    return (c, length)
```
Comment (TomDonoghue) [inline fooof/synth.py]: This return doesn't need to return a tuple, right? It could be: return c, length?

**Comment by TomDonoghue** [comment]
@ArcadeShrimp : okay, I left a bunch of review comments in the code, mostly for style, documentation and documentation. Go through and update these things in the code. 

Code - we need tests!
Add at least one test case to fooof/tests/test_synth. You can mimic `test_param_samples` - just add something really basic to test param_iter executes, and seems to do what it's supposed to.

**Comment by TomDonoghue** [comment]
Code Update:

If you don't feed in a np.array, it fails in a sorta confusing way - with an UnboundLocal error, because 'ind' ends up not defined. We should fix this up - and I think the way it currently checks for np.arrays as a way to find the iterated parameter perhaps not the best way to this, overall. 

Pulling together everything we've learned from this process, I suggest one final update to the procedure:

Instead of passing in something like np.arange(8, 12, 1), lets use this custom object:
```
class Stepper():
    
    def __init__(self, lst):
        
        dat = np.arange(*lst)
        self._len = len(dat)
        self.data = iter(dat)
    
    def __len__(self):
        return self._len
        
    def __next__(self):
        return next(self.data)

    def __iter__(self):
        return self.data
```

This will change the use, a bit, to something like:
```
cf_step = Stepper([8, 12, 1])
osc_params = param_iter([cf_steps, 1, 1])
gen_group_power_spectra(len(cf_step), [3, 40], [1, 1], osc_params) 
```
^I've checked, and this works. Does this look like a good idea to you?

File: fooof/synth.py [hunk 2]
```diff
@@ -39,7 +104,7 @@ def param_sampler(params, probs=None):
         params = [_check_flat(lst) for lst in params]
 
     # In order to use numpy's choice, with probabilities, choices are made on indices
-    #  This is because the params can be a messy-sized list, that numpy choice does not like
+    #  This is gbecause the params can be a messy-sized list, that numpy choice does not like
```
Comment (TomDonoghue) [inline fooof/synth.py]: I think you added a typo here by mistake - to fix!

**Comment by TomDonoghue** [comment]
Last thing: the version you have doesn't deal with multiple oscillations, right? I couldn't get it to work for something like:

```
osc_1 = [np.arange(8, 12, 1), 1, 1]
osc_2 = [20, 1, 1]
osc_params = param_iter([osc_1, osc_2])
```

Am I using it wrong, or is this just not supported yet? If it doesn't do this yet (that's totally fine!), my suggestion above, as is, also doesn't yet support this - but we will need it to be able to do this for some band ratios stuff: so let's update everything from above, and then consider what it needs to support multiple oscillations.

File: fooof/synth.py (hunk 1 above)
Comment (ArcadeShrimp) [reply fooof/synth.py]: isnt the line:
return c, length 
an implicit tuple? 

https://www.geeksforgeeks.org/g-fact-41-multiple-return-values-in-python/

**Comment by ArcadeShrimp** [comment]
Should the Stepper class be defined inside synth.py?

File: fooof/tests/test_synth.py [hunk 3]
```diff
@@ -7,6 +8,27 @@
...
 ###################################################################################################
 
+def test_param_iter():
+
+    # Test list of params
+    step = Stepper([8, 12, .1])
+    osc = [1, 2, 3, step, 5]
+    iter_1 = param_iter(osc)
+
+    for i in range(len(step)):
```
Comment (TomDonoghue) [inline fooof/tests/test_synth.py]: Drop range:
`for val in step` or `for ind, val in enumerate(step)`:

File: fooof/tests/test_synth.py [hunk 4]
```diff
@@ -7,6 +8,27 @@
 ###################################################################################################
 ###################################################################################################
 
+def test_param_iter():
+
```
Comment (TomDonoghue) [inline fooof/tests/test_synth.py]: Add a simple 2 value (reflect AP) and 3 value (reflects osc) test case. 

Add comment labels:
`# Tests 2 value (used for AP generation)`

`# Test 3 value (used for single osc generation)`

`# Test n value (test generalization)`

`# Test multiple oscillations case`

File: fooof/synth.py [hunk 5]
```diff
@@ -17,6 +17,96 @@
 
 SynParams = namedtuple('SynParams', ['background_params', 'gaussian_params', 'nlv'])
 
+class Stepper():
```
Comment (TomDonoghue) [inline fooof/synth.py]: Add a class methods to check Stepper values:

Note: use a static method, attached to the Stepper object.
```
@staticmethod
def _check_values(start, stop, step):
    if not start < stop:
        raise ValueError()
    if not step < (stop - start):
        raise ValueError()
```

**Comment by TomDonoghue** [comment]
We were failing on Python3.7, which appears to be due to a change in how StopIteration is handled:
https://www.python.org/dev/peps/pep-0479/

Based on what's written there, it seems we have to have explicit handling of the StopIteration in param_iter, or it will raise a RuntimeError in py37. Adding this (using a try/except to catch the StopIteration and in param_iter, and return None when we hit it - a more explicit end-of-generator procedure), seems to fix this.

**Pull request merged**